.github/workflows/ci.yml

```diff
@@ -0,0 +1,197 @@
+name: CI
+
+on:
+    pull_request:
+    push:
+        branches:
+            - main
+
+concurrency:
+    group: ci-${{ github.workflow }}-${{ github.event.pull_request.number || github.ref }}
+    cancel-in-progress: true
+
+jobs:
+    build_and_test_nix:
+        name: Build and test
+        timeout-minutes: 30
+        runs-on: ${{ matrix.runner }}
+        strategy:
+            fail-fast: false
+            matrix:
+                name: [ubuntu-latest, macOS-arm-latest]
+                rust: [stable]
+                include:
+                    - name: ubuntu-latest
+                      os: ubuntu-latest
+                      release-os: linux
+                      release-arch: amd64
+                      runner: [self-hosted, linux, X64]
+                    - name: macOS-arm-latest
+                      os: macOS-latest
+                      release-os: darwin
+                      release-arch: aarch64
+                      runner: [self-hosted, macOS, ARM64]
+        steps:
+            - name: checkout
+              uses: actions/checkout@v4
+
+            - name: Install ${{ matrix.rust }} rust
+              uses: dtolnay/rust-toolchain@master
+              with:
+                toolchain: ${{ matrix.rust }}
+            - name: install cargo-nextest
+              uses: taiki-e/install-action@v2
+              with:
+                tool: nextest
+            - name: run tests
+              run: |
+                cargo nextest run --all-features --lib --bins --tests --no-fail-fast
+              env:
+                RUST_LOG: "TRACE"
+            - name: build
+              run: |
+                cargo build --release
+
+    build_and_test_windows:
```

Comment:
we could remove windows for now I'd say if we want to save on runner time, don't think people will run this on windows